Add torch.ops per overload API

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72673
* __->__ #72206

Updated `torch.ops` API to return an `OpOverloadPacket` instead of a built-in method python object. 
Added a check to disallow 'default' as an overload name.
 
Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

anjali411 committed:
Add torch.ops per overload API

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.ops per overload API"




[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33955635).

albanD reviewed (approved):
🚀

anjali411 committed:
Update on "Add torch.ops per overload API"


Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.ops per overload API"


Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.ops per overload API"


Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

[ghstack-poisoned]

albanD commented on test/jit/test_custom_operators.py:
```diff
@@ -78,14 +74,6 @@ def test_passing_one_positional_but_not_the_second(self):
         ):
             torch.ops.aten.type_as(torch.ones(5, 5))
 
-    def test_passing_an_argument_both_as_positional_and_kwarg(self):
```
Why are these tests removed?

anjali411 replied:
bc breaking change

anjali411 committed:
Update on "Add torch.ops per overload API"


Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.ops per overload API"


Updated `torch.ops` API to return an `OpOverloadPacket` instead of a built-in method python object. 
Added a check to disallow 'default' as an overload name.
 
Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.ops per overload API"


Updated `torch.ops` API to return an `OpOverloadPacket` instead of a built-in method python object. 
Added a check to disallow 'default' as an overload name.
 
Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33955635).

albanD commented on torch/_ops.py:
```diff
@@ -99,39 +99,59 @@ def __getattr__(self, key):
+        # `_get_operation_overload` (which is an expensive operation).
+        # This is done to prevent any potential slowdown. This list can be extended
+        # if there exists other attributes like `__name__` that only exist on self._op and not on the
+        # opoverloadpacket.
```
nit: I think we can mention here that this is ok as aten ops cannot start with "__" ever.
Also since we are guaranteed that anything that starts with "__" isn't an op, not sure we need the "endswith" check?

albanD commented on torch/_ops.py:
```diff
@@ -99,39 +99,59 @@ def __getattr__(self, key):
+            # throw an attribute error with a message containing
+            # an object name different from the one the attribute
+            # query was performed on.
+            raise AttributeError("'{}' object has no attribute '{}'".format(str(self), key)) from None
```
Why do you set "from None" ?

anjali411 replied:
I had nested try except before so needed it but don't need it anymore

anjali411 replied:
without from None:
```
>>> torch.ops.aten.add.__bla
Traceback (most recent call last):
  File "/fsx/users/chourdiaanjali/pytorch/torch/_ops.py", line 111, in __getattr__
    return getattr(self._op, key)
AttributeError: 'builtin_function_or_method' object has no attribute '__bla'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/fsx/users/chourdiaanjali/pytorch/torch/_ops.py", line 117, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(str(self._op), key))
AttributeError: '<built-in method add of PyCapsule object at 0x7f4b01952060>' object has no attribute '__bla'
```

with it:
```
>>> torch.ops.aten.add.__bla
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/fsx/users/chourdiaanjali/pytorch/torch/_ops.py", line 117, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(str(self._op), key)) from None
AttributeError: '<built-in method add of PyCapsule object at 0x7fc270924060>' object has no attribute '__bla'
```

anjali411 replied:
come to think of it, maybe I don't need a try except here

albanD replied:
If we make the re-throw message better, I guess we can keep the from None.

albanD commented on torch/_ops.py:
```diff
@@ -99,39 +99,59 @@ def __getattr__(self, key):
-            except AttributeError:
-                raise AttributeError("'{}' object has no attribute '{}'".format(str(self), key)) from None
+            raise AttributeError(
+                "'{}' object has no attribute '{}'".format(str(self), key)
```
Should we be clearer here that we failed to get an op overload for a given op and a given overload?

anjali411 committed:
Update on "Add torch.ops per overload API"


Updated `torch.ops` API to return an `OpOverloadPacket` instead of a built-in method python object. 
Added a check to disallow 'default' as an overload name.
 
Differential Revision: [D33955635](https://our.internmc.facebook.com/intern/diff/D33955635)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33955635).

albanD reviewed (approved):
SGTM
